src/parser/parser.cpp

```diff
@@ -138,39 +137,78 @@ void Parser::ParseQuery(const string &query) {
+				// if it succeeded, we transform the Postgres parse tree into a list of
+				// SQLStatements
+				transformer.TransformParseTree(parser.parse_tree, statements);
+			} else {
+				parsing_failed = true;
+			}
 		}
-	}
-	if (!parser_error.empty()) {
-		if (options.extensions) {
-			for (auto &ext : *options.extensions) {
-				D_ASSERT(ext.parse_function);
-				auto result = ext.parse_function(ext.parser_info.get(), query);
-				if (result.type == ParserExtensionResultType::PARSE_SUCCESSFUL) {
-					auto statement = make_uniq<ExtensionStatement>(ext, std::move(result.parse_data));
-					statement->stmt_length = query.size();
-					statement->stmt_location = 0;
-					statements.push_back(std::move(statement));
-					return;
+		// if DuckDB fails to parse the entire sql string, break the string down into individual statements
+		// so that parser extensions can help parse some statements
+		if (parsing_failed) {
+			vector<string> query_statements;
+			auto tokens = Tokenize(query);
+			auto next_statement_start = 0;
+			for (size_t i = 1; i < tokens.size(); ++i) {
+				auto &t_prev = tokens[i - 1];
+				auto &t = tokens[i];
+				if (t_prev.type == SimplifiedTokenType::SIMPLIFIED_TOKEN_OPERATOR) {
+					if (StringUtil::Contains(query.substr(t_prev.start, t.start - t_prev.start), ";")) {
+						query_statements.emplace_back(
+						    query.substr(next_statement_start, t.start - next_statement_start));
+						next_statement_start = tokens[i].start;
+					}
 				}
-				if (result.type == ParserExtensionResultType::DISPLAY_EXTENSION_ERROR) {
-					throw ParserException(result.error);
+			}
+			query_statements.emplace_back(query.substr(next_statement_start, query.size() - next_statement_start));
+			for (auto const &s : query_statements) {
+				PostgresParser another_parser;
+				another_parser.Parse(s);
+				// first see if DuckDB can parse this individual query statement
+				if (another_parser.success) {
+					if (!another_parser.parse_tree) {
+						// empty statement
+						return;
+					}
+
+					transformer.TransformParseTree(another_parser.parse_tree, statements);
+				} else {
+					// let the extension parse the statement which DuckDB failed to parse
+					if (options.extensions) {
+						for (auto &ext : *options.extensions) {
+							D_ASSERT(ext.parse_function);
+							auto result = ext.parse_function(ext.parser_info.get(), s);
+							if (result.type == ParserExtensionResultType::PARSE_SUCCESSFUL) {
+								auto statement = make_uniq<ExtensionStatement>(ext, std::move(result.parse_data));
+								statement->stmt_length = s.size();
+								statement->stmt_location = 0;
+								statements.push_back(std::move(statement));
+							}
+							if (result.type == ParserExtensionResultType::DISPLAY_EXTENSION_ERROR) {
+								throw ParserException(result.error);
+							} else if (result.type == ParserExtensionResultType::DISPLAY_ORIGINAL_ERROR) {
+								parser_error = QueryErrorContext::Format(query, another_parser.error_message,
```

Comment:
The offset and query will be off in the parser error produced here (it will refer to the individual statement instead of the original query). This might be desirable, calling out for discussion. 

We should add some test cases around this.